File: C#/ConverterApp/Converter.cs
using System;

namespace ConverterApp
{
    public class Converter<TInput, TOutput>
    {
        private Func<TInput, TOutput> converterFunction;

        public Converter(Func<TInput, TOutput> converterFunction)
        {
            this.converterFunction = converterFunction;
        }

        public TOutput Convert(TInput input)
        {
            return converterFunction(input);
        }
    }
}

File: C++/ConverterApp/converter.cpp
#include "converter.h"

template <typename TInput, typename TOutput>
Converter<TInput, TOutput>::Converter(std::function<TOutput(TInput)> func) : func(func) {}

template <typename TInput, typename TOutput>
TOutput Converter<TInput, TOutput>::Convert(TInput input) const {
    return func(input);
}

template class Converter<int, std::string>;
template class Converter<std::string, int>;